sdk/core/Azure.Core/tests/BearerTokenAuthenticationPolicyTests.cs

```diff
@@ -789,6 +789,109 @@ public async Task BearerTokenAuthenticationPolicy_BackgroundRefreshFailsAndLogs(
             Assert.IsTrue(logged);
         }
 
+        [Test]
+        public async Task BearerTokenAuthenticationPolicy_SwitchedTenants()
+        {
+            var responses = new[]
+            {
+                new MockResponse(401)
+                    .WithHeader("WWW-Authenticate", @"Bearer authorization=""https://login.windows.net/de763a21-49f7-4b08-a8e1-52c8fbc103b4"", resource=""https://vault.azure.net"""),
+
+                new MockResponse(200),
+                new MockResponse(200),
+
+                // Moved tenants.
+                new MockResponse(401)
+                    .WithHeader("WWW-Authenticate", @"Bearer authorization=""https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47"", resource=""https://vault.azure.net""")
+                    .WithJson("""
+                    {
+                        "error": {
+                            "code": "Unauthorized",
+                            "message": "AKV10032: Invalid issuer. Expected one of https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/, https://sts.windows.net/f8cdef31-a31e-4b4a-93e4-5f571e91255a/, https://sts.windows.net/e2d54eb5-3869-4f70-8578-dee5fc7331f4/, https://sts.windows.net/33e01921-4d64-4f8c-a055-5bdaffd5e33d/, https://sts.windows.net/975f013f-7f24-47e8-a7d3-abc4752bf346/, found https://sts.windows.net/96be4b7a-defb-4dc2-a31f-49ee6145d5ab/."
+                        }
+                    }
+                    """),
+
+                new MockResponse(200),
+            };
+
+            var transport = CreateMockTransport(responses);
+
+            var credential = new TokenCredentialStub((r, c) => new(Guid.NewGuid().ToString(), DateTimeOffset.Now.AddHours(2)), IsAsync);
+            var policy = new ChallengeBasedAuthenticationPolicy(credential, "scope");
+
+            await SendGetRequest(transport, policy, uri: new("https://example.com/1/Original"));
+            Assert.AreEqual("de763a21-49f7-4b08-a8e1-52c8fbc103b4", policy.TenantId);
+
+            await SendGetRequest(transport, policy, uri: new("https://example.com/1/Original"));
+            Assert.AreEqual("de763a21-49f7-4b08-a8e1-52c8fbc103b4", policy.TenantId);
+
+            await SendGetRequest(transport, policy, uri: new("https://example.com/1/Original"));
+            Assert.AreEqual("72f988bf-86f1-41af-91ab-2d7cd011db47", policy.TenantId);
```

Comment:
Is there a way to validate that the new token is generated when the tenant changes?

Reply:
I can see, but at some point I'm just going to end up reimplementing Key Vault's `ChallengeBasedAuthenticationPolicy`, which already tests this scenario. Maybe I should just copy it and be done with it rather than try to come up with a minimal repro.